solr/core/src/test/org/apache/solr/search/TestJoinQueryPerformance.java

```diff
@@ -0,0 +1,235 @@
+    }
+
+    return words;
+  }
+
+
+  @Test
+  public void testJoinPerformanceAsMainQueryHitsIncrease() throws Exception {
+    final String joinQueryBase = "{!join fromIndex=" + FROM_COLLECTION_NAME + " from=permissions_ss to=acl_ss cache=false";
+    final String fromQuery = "userid_s:user25"; // The higher the user number, the more permissions he has attached to his name (1-100)
+    final String standardJoin = joinQueryBase + "}" + fromQuery;
+    final String noScoreJoin = joinQueryBase + " score=none}" + fromQuery;
+    final String tpiJoin = joinQueryBase + " toplevel=true}" + fromQuery;
+
+    try (HttpSolrClient client = new HttpSolrClient.Builder("http://localhost:8983/solr").build()) {
+      for ( int i = 0; i < VAL_MAX; i+=20) {
+        final String mainQuery = "val_i:[1 TO " + (i+1) + "]";
+        final QueryResponse standardJoinRsp = client.query(TO_COLLECTION_NAME, new SolrQuery("q", mainQuery, "fq", standardJoin), SolrRequest.METHOD.POST);
+        final QueryResponse tpiJoinRsp = client.query(TO_COLLECTION_NAME, new SolrQuery("q", mainQuery, "fq", tpiJoin), SolrRequest.METHOD.POST);
+        final QueryResponse noScoreJoinRsp = client.query(TO_COLLECTION_NAME, new SolrQuery("q", mainQuery, "fq", noScoreJoin), SolrRequest.METHOD.POST);
+        final long numFound = tpiJoinRsp.getResults().getNumFound();
+
+        System.out.println(i + "," + numFound + "," + standardJoinRsp.getQTime() + "," + noScoreJoinRsp.getQTime() + "," + tpiJoinRsp.getQTime());
+      }
+    }
+  }
+
+  @Test
+  public void testJoinPerformanceAsFromQueryHitsIncrease() throws Exception {
+    final String mainQuery = "val_i:[1 TO 250]"; // Half the docs match the query (250K)
+
+    final String joinQueryBase = "{!join fromIndex=" + FROM_COLLECTION_NAME + " from=permissions_ss to=acl_ss cache=false";
+
+
+    try (HttpSolrClient client = new HttpSolrClient.Builder("http://localhost:8983/solr").build()) {
+      for ( int i = 1; i <= USER_MAX; i++) {
+        final String fromQuery = "userid_s:user" + i;
+        final String standardJoin = joinQueryBase + "}" + fromQuery;
+        final String noScoreJoin = joinQueryBase + " score=none}" + fromQuery;
+        final String tpiJoin = joinQueryBase + " toplevel=true}" + fromQuery;
+
+        final QueryResponse standardJoinRsp = client.query(TO_COLLECTION_NAME, new SolrQuery("q", mainQuery, "fq", standardJoin), SolrRequest.METHOD.POST);
+        final QueryResponse tpiJoinRsp = client.query(TO_COLLECTION_NAME, new SolrQuery("q", mainQuery, "fq", tpiJoin), SolrRequest.METHOD.POST);
+        final QueryResponse noScoreJoinRsp = client.query(TO_COLLECTION_NAME, new SolrQuery("q", mainQuery, "fq", noScoreJoin), SolrRequest.METHOD.POST);
+        final long numFound = tpiJoinRsp.getResults().getNumFound();
+
+        System.out.println(i + "," + numFound + "," + standardJoinRsp.getQTime() + "," + noScoreJoinRsp.getQTime() + "," + tpiJoinRsp.getQTime());
```

Comment:
Printing to stdout is prosecuted.

Reply:
Hey Mikhail, thanks for the review.  I have no intentions of committing this file.  I kept it in the PR up to this point so that people could reproduce/tweak the perf test I ran.  But I'm going to remove it prior to commit.

(In fact, I thought I already had removed it, but there was one last "cleanup" commit I'd neglected to push until right now.  You should see this file gone from the PR now.)